cdc/generation.cc

```diff
@@ -123,176 +123,110 @@ const std::vector<token_range_description>& topology_description::entries() cons
     return _entries;
 }
 
-static stream_id make_random_stream_id() {
+static stream_id create_stream_id(dht::token t) {
     static thread_local std::mt19937_64 rand_gen(std::random_device().operator()());
     static thread_local std::uniform_int_distribution<int64_t> rand_dist(std::numeric_limits<int64_t>::min());
 
-    return {rand_dist(rand_gen), rand_dist(rand_gen)};
+    return {dht::token::to_int64(t), rand_dist(rand_gen)};
 }
 
-/* Given:
- * 1. a set of tokens which split the token ring into token ranges (vnodes),
- * 2. information on how each token range is distributed among its owning node's shards
- * this function tries to generate a set of CDC stream identifiers such that for each
- * shard and vnode pair there exists a stream whose token falls into this
- * vnode and is owned by this shard.
- *
- * It then builds a cdc::topology_description which maps tokens to these
- * found stream identifiers, such that if token T is owned by shard S in vnode V,
- * it gets mapped to the stream identifier generated for (S, V).
- */
-// Run in seastar::async context.
-topology_description generate_topology_description(
-        const db::config& cfg,
-        const std::unordered_set<dht::token>& bootstrap_tokens,
-        const locator::token_metadata& token_metadata,
-        const gms::gossiper& gossiper) {
-    if (bootstrap_tokens.empty()) {
-        throw std::runtime_error(
-                "cdc: bootstrap tokens is empty in generate_topology_description");
+class topology_description_generator final {
+    const db::config& _cfg;
+    const std::unordered_set<dht::token>& _bootstrap_tokens;
+    const locator::token_metadata& _token_metadata;
+    const gms::gossiper& _gossiper;
+
+    // Compute a set of tokens that split the token ring into vnodes
+    auto get_tokens() const {
+        auto tokens = _token_metadata.sorted_tokens();
```

Comment:
How does this work with multiple data centers and racks? A vnode with multiple data centers and racks ignores adjacent tokens that are on the same rack or same dc.

I think I found a problem: let's say two adjacent tokens are (300001, 300002]. Then the code to find the first token for a shard will fail for all but one shard (the one that olds 300002). But if the token 300001 is held by a node on the same rack, or a different data center, then the beginning of the vnode is not 300001 but somewhere before, let's say 100001. Now you can find shards for all the vnode range.

Reply:
Aren't different datacenters having separate token rings?

Reply:
I don't know how racks affect the ring - will check it out.

Reply:
@kbr- Have you considered multi DC and racks when you were doing the initial code? Could you comment, please?

Reply:
No, there's one big ugly ring and the replication strategy ignores racks in the same dc and nodes in different dc:s.

Reply:
> @kbr- Have you considered multi DC and racks when you were doing the initial code? Could you comment, please?

Tokens and DCs/Racks are orthogonal concepts. We're talking apples and oranges, so there's really nothing to consider here.

> let's say two adjacent tokens are (300001, 300002]. Then the code to find the first token for a shard will fail for all but one shard (the one that olds 300002). But if the token 300001 is held by a node on the same rack, or a different data center, then the beginning of the vnode is not 300001 but somewhere before, let's say 100001.

Each node chooses a set of 255 (by default) tokens. These tokens are **chosen** by this node, and this node is the **owner** of these tokens.

The token ring is the sum of chosen tokens for all nodes. There is 255 * num_nodes chosen tokens in total.

A vnode is a smallest token range `(x, y]` such that `x` and `y` are chosen. "smallest" means that there is no chosen token in between.

If 300001 and 300002 are chosen, then `(300001, 300002]` is a vnode, and `(100001, 300002]` is not a vnode.

The definition of vnode does not include "DC" or "rack" anywhere.

A replication strategy iterates over all vnodes in the token ring in the clockwise direction -- or equivalently, it iterates over all chosen tokens in the clockwise direction -- and looks at the owner of the end token of this vnode (for vnode `(x, y]`, it looks at the owner of `y`). It doesn't look at tokens `z` such that `x < z <= y`, because they give no information: the concept of "owner" is not defined for these tokens. In the case of NTS, it will sometimes decide to not include an owner in the calculated replica set if it already included an owner in the same DC/rack.

For two tokens in the same vnode, the set of calculated replicas will be the same, because the sequence of chosen tokens "to the right" of these tokens is the same. Intuitively, the replication strategy (any replication strategy) cannot distinguish between two tokens in the same vnode.

We choose streams so that every vnode has at least one stream whose token lies in this vnode. When we consider a partition key (when making a write), we take its token and we take a stream whose token lies in the same vnode as the token of the key. Because they lie in the same vnode, the replica set for both of them will be the same.

Reply:
Also I made this document

https://docs.google.com/document/d/1_p5mJUSWb_CoCD9Fg7mum2KH893FuvM2uRquQrXZmGY/edit?usp=sharing

with a detailed explanation of tokens, token ring, replication strategies, and how these concepts play with each other.

(it's not published yet because it's part of a bigger document that I'm working on from time to time which is not finished yet)

Reply:
@avikivity we can't do the optimization you mentioned. We need stream for exactly each vnode because writes will be going to all replicas (in all datacenters).

Reply:
I think @kbr- is correct. But let's make sure. Where is the code that uses the narrow definition of vnode (two adjacent tokens) rather than the wide definition (two tokens that define a range according to the replication strategy)?

Reply:
For NTS, this is the loop that walks over the vnodes (their end tokens) and decides whether or not to take an owner of this token into the set of replicas:
```
    for (auto& next : tm.ring_range(search_token)) {
        if (dcs_to_fill == 0) {
            break;
        }

        inet_address ep = *tm.get_endpoint(next);
        auto& loc = tp.get_location(ep);
        auto i = dcs.find(loc.dc);
        if (i != dcs.end() && i->second.add_endpoint_and_check_if_done(ep, loc)) {
            --dcs_to_fill;
        }
    }
```
(in `network_topology_strategy::calculate_natural_replicas`).

`tm` is token_metadata. `tm.ring_range(search_token)` takes the first chosen token "equal or to the right, clockwise" of `search_token`. So `next` walks over chosen tokens in clockwise direction, and decides whether or not we should include its owner to the replica set.

`ring_range` is implemented using `token_iterator_impl` in token_metadata, which uses `tm.sorted_tokens()`. It doesn't look at DC/rack placement.

The important fact for us is that for two values of `search_token` for which `tm.ring_range(search_token)` will be the same, the set of calculated replicas will be the same.

In other words, for two tokens that are in the same vnode, the set of calculated replicas will be the same.

For SimpleStrategy we have a similar, but simpler loop:
```
    for (auto& token : tm.ring_range(t)) {
        if (endpoints.size() == replicas) {
           break;
        }
        auto ep = tm.get_endpoint(token);
        assert(ep);

        endpoints.push_back(*ep);
    }
```
(in `simple_strategy::calculate_natural_endpoints`).

Reply:
To complete my understanding, where does cdc make the connection between the mutation's token and the chosen stream?

Reply:
cdc/log.cc:
```
    std::tuple<mutation, stats::part_type_set> transform(const mutation& m, const cql3::untyped_result_set* rs, api::timestamp_type ts, bytes tuuid, int& batch_no) {
        auto stream_id = _ctx._cdc_metadata.get_stream(ts, m.token());
```

Where `get_stream` is a method of `cdc_metadata`, which is defined in cdc/metadata.hh.

And this is the implementation of `get_stream`:
```
static cdc::stream_id get_stream(
        const cdc::token_range_description& entry,
        dht::token tok) {
    // The ith stream is the stream for the ith shard.
    auto shard_cnt = entry.streams.size();
    auto shard_id = dht::shard_of(shard_cnt, entry.sharding_ignore_msb, tok);

    if (shard_id >= shard_cnt) {
        on_internal_error(cdc_log, "get_stream: shard_id out of bounds");
    }

    return entry.streams[shard_id];
}

static cdc::stream_id get_stream(
        const std::vector<cdc::token_range_description>& entries,
        dht::token tok) {
    if (entries.empty()) {
        on_internal_error(cdc_log, "get_stream: entries empty");
    }

    auto it = std::lower_bound(entries.begin(), entries.end(), tok,
            [] (const cdc::token_range_description& e, dht::token t) { return e.token_range_end < t; });
    if (it == entries.end()) {
        it = entries.begin();
    }

    return get_stream(*it, tok);
}
```

Reply:
Thanks for the detailed explanations.